text/0000-array-private-classes.md

```diff
@@ -0,0 +1,366 @@
+general type that fullfils the contract of the function (in the case of the
+functions discussed in this RFC, iteration).
+- Collections' functions `keys`, `values` and `pairs` definitions are made more
+general. Iterators implementation details are not public. Internal classes used
+by implementation like `*Keys`, `*Values` and `*Pairs` are now
+[opaque data types](https://en.wikipedia.org/wiki/Opaque_data_type). Generally,
+when using these collection classes, clients are not interested by the iterators
+implementation, but by the types these iterators return and that is provided by
+the generic parameters.
+- The generic return signature of these 3 iterating functions is simpler to
+understand for clients of collection classes.
+- Reduces the number of public classes in the standard library by hiding 18
+specialised classes (iterators implementations) of which 3 are from the
+`builtin` module.
+- The interface `CollectionIterator` is added to create rewindable iterators
+(can be re-start from first value).
+
+This change remains compatible with the existing code base but for client code
+that is directly using the classes `*Keys`, `*Values` and `*Pairs`. A search on
+Github shows that the impact is very limited.
+
+To quote Antoine de Saint-Exupéry:
+
+> Perfection is achieved, not when there is nothing more to add, but when there
+> is nothing left to take away.
+
+Currently the API makes a guarantee that the return value of each of these functions
+is a specific concrete class with a specific public name. That particular guarantee
+does not actually serve any practical needs of users (except, arguably, the virtual
+calls caveat that was mentioned on [Github discussion](https://github.com/ponylang/rfcs/pull/193#issuecomment-1009733064)). If it is not helpful to users in practice,
+it can be removed. And if it can be removed, then it should be removed, in the
+name of taking another step toward stabilizing the Pony standard library's API.
+
+It's worth mentioning that if this change was being debated in reverse (the
+current standard library had minimal interfaces for the return values, and
+somebody proposed to make them public concrete classes instead, because they
+found some tangible benefit in doing so), then it would not be a breaking change
+at all, because the new proposed type would be a subtype of the existing return
+types.
+
+It'll always be easier to get more specific with these return types than more
+general. So we should default to offering the most general return type that is
+still useful to users, and ratchet down to more specific subtypes as needed (as
+tangible use cases for more specific return types are discovered). Ratcheting down
+will always be possible without a breaking change, but ratcheting up will be a
+breaking change each time. It's best to make such "ratchet up" breaking changes
+like this RFC before we reach Pony 1.0.0 as part of a general effort to make Pony
+and its standard library stable.
```

Comment:
I do not agree with this as motivation. It is always easier to start with many concrete implementations/returns which are then used to derive a generic implementation/return and trying to be generic before concrete presents more room for instability.